### ISSUE TITLE
Progressive Web App Functionality

### ISSUE THREAD

**Issue by tayambamwanza** [angular-checklist/angular-checklist#15]
It would be great to have this as an installable web app, I can implement this for you if you want.

- [x] Create dev branch

- [x] Create dev firebase project for a test hosting site "angular-checklist-dev"?

- [x] ng add @angular/pwa

- [x] add service worker to app.component.ts

- [x] automatic update logic

- [x] manual update logic

- [ ] manual update button

**Comment by KwintenP** [comment]
Hi, 

This is definitely something that's on our mind. I wanna be really careful here though as I know there have been some issues before with PWA in Angular.

I want to make sure that people always get the latest version if new items are added. And I've had first hand experience with this not working for some project that I did. 

Don't know if the benefit of having a PWA is worth the potential errors it brings 🤔. Or am I being to conservative here @d3lm?

**Comment by tayambamwanza** [comment]
@KwintenP, that is true, I had these issues myself but this has pretty much been the story of Angular as well :laughing: initially it was rough around the edges and had some inconsistencies but it's matured very nicely and I've managed to use it in a fashion that's reliable in my latest project which is why I'm volunteering to do it.

You can make a new project called dev-angular-checklist on firebase hosting and deploy the app there temporarily to test, then you don't risk affecting the main site.

**Comment by d3lm** [comment]
@tayambamwanza Hey there, thanks for showing interest in contributing to this project!

As @KwintenP said, we were also thinking about making the Angular Checklist a PWA but I can also agree that it's still rough around the edges and from first hand experience I can tell that there are still a bunch of issues.

I am not sure how beneficial it would be for the checklist. I do see the benefit of the being available offline but we'd need to make sure that the version the user has is **always** up-to-date as the content is really important and I don't want people to stay on old versions just because the SW is not updating properly.

So if we can figure out how to deal with this and implement it in a reliable way, I'd be up for this! I think the main focus shouldn't be mobile here but rather offline availability. But we have to keep in mind, that at some point we want to integrate Firebase and sync the data to a database.

@KwintenP, you're definitely not too conservative about this. It's just that we both had to face several issues in the past. Also I don't want the UX to suffer. The app was build to provide a great UX and I want to keep it that way.

Again, if we can find a way to make this work reliably, sure, why not.

@KwintenP Thoughts?

**Comment by d3lm** [comment]
Ok, so I think we should try it out in a dev branch and see how it's working. But if we have a proper update mechanism in place it shouldn't be a problem. 

It would be really nice to have this offline as well cause if you're on a train or on a plane for example, you can still work with the checklist if you want.

**Comment by tayambamwanza** [comment]
Thanks for the responses, I'd be willing to help out with setting this up, so let me know if you'd be willing to see what I come up with after you set up the dev branch & dev firebase project.

One of the fail-safes that I implemented was having an actual button to "check for updates" somewhere in my app (normally at the bottom of the sidenav) so it would force the service worker to get the latest version if the automatic update failed somehow.

**Comment by d3lm** [comment]
@tayambamwanza I like that safety net of having a button for the worst case. Cause one problem with PWA's is that if you forget to build in an interval that is polling for updates, users are stuck for a cached version of the app. One can seriously screw up here 😆 But having that button sounds like a really good idea.

**Comment by d3lm** [comment]
Yep agreed @KwintenP. If we get this up and running and that users will **always** be on the latest version and update if their is a new version available, definitely!

**Comment by tayambamwanza** [comment]
Alright great then, i'll be on standby i'll make a checklist if you don't mind :)

**Comment by d3lm** [comment]
I will go ahead and set up the `dev` branch. @KwintenP will then take care of setting up a Firebase project that is used to deploy the branch.

**Comment by KwintenP** [comment]
I'll try to set this up asap.

**Comment by KwintenP** [comment]
Misclicked 🙈

**Comment by tayambamwanza** [comment]
@d3lm , I'm almost setup but when I try to build I get this error:

`ERROR in ./src/app/checklist/state/checklist.reducer.ts
Module not found: Error: Can't resolve '../../../assets/content.json' in '~/angular-checklist/src/app/checklist/state'`

**Comment by KwintenP** [comment]
Also reachable from https://dev.angular-checklist.io/projects from now on (certificate is still generating on firebase hosting).

**Comment by d3lm** [comment]
@tayambamwanza Did you run `yarn start`? It seems that you only ran the app but not the script that builds the content. The `start` script does both.

**Comment by d3lm** [comment]
Any progress update on this one?

**Comment by tayambamwanza** [comment]
Not yet, I've been away but back now so I'll start work on it soon

**Comment by d3lm** [comment]
No worries! I just wanted to check for any updates to make sure this is still being worked on and not stale.

**Comment by tayambamwanza** [comment]
Don't worry about it hasn't gone stale :smile:

**Comment by tayambamwanza** [comment]
I installed the @angular/pwa package without making changes. I need icons now, would you be able to send me something?

![screenshot from 2019-01-11 19-21-19](https://user-images.githubusercontent.com/11971384/51049330-3b161b80-15d6-11e9-9cf1-5c27767d5b21.png)
![screenshot from 2019-01-11 19-21-54](https://user-images.githubusercontent.com/11971384/51049333-3d787580-15d6-11e9-88df-72d2e369fc10.png)

**Comment by d3lm** [comment]
Yep I can generate those for you. Will do that next week.

**Comment by tayambamwanza** [comment]
Fantastic, much appreciated

**Comment by tayambamwanza** [comment]
@d3lm Thanks haha, didn't realise you'd send it already, appreciate it.

**Comment by tayambamwanza** [comment]
@d3lm Quick question: Should I include the meta tags for iOS Safari or just support android?

I normally use this tool to generate the iOS specific icons with the appropriate names and html tags [Favicon Generator](https://realfavicongenerator.net/)

I just did a quick google and found this article as well [PWA Web App Manifest support for iOS!](https://medium.com/@aganglada/pwa-web-app-manifest-support-for-ios-fedd66d0ee88)

**Comment by d3lm** [comment]
I'd say that we support both iOS and Android as those have a huge market coverage.

**Comment by tayambamwanza** [comment]
@d3lm @KwintenP Apologies for the delay I submitted a pull request for this just now, things got a bit hectic and still are but I thought I'd try sort this out. Any adjustments and suggestions are welcome.

**Comment by GregorBiswanger** [comment]
What is the current status on this? As a PWA app would be great!
